ds701_book/27-RNN.qmd

```diff
@@ -0,0 +1,640 @@
+    train_losses.append(train_loss)
+
+# Plot the training losses over epochs
+plt.figure(figsize=(10, 5))
+plt.plot(range(1, epochs + 1), train_losses, marker='o', linestyle='-', color='b')
+plt.title('Training Loss over Epochs')
+plt.xlabel('Epochs')
+plt.ylabel('Loss')
+plt.xticks(range(1, epochs + 1))  # Set x-tick marks to be integers
+plt.grid(True)
+plt.show()
+```
+
+## Evaluation
+
+Let's evaluate the model's performance.
+
+```{python}
+# Evaluate the model
+model.eval()
+predictions = []
+trues = []
+with torch.no_grad():
+    for X, y in test_loader:
+        preds = model(X)
+        predictions.extend(preds.numpy())
+        trues.extend(y.numpy())
+
+# Rescale predictions and true to original scale
+predictions_rescaled = scaler.inverse_transform(predictions)
+true_rescaled = scaler.inverse_transform(trues)
+
+# Plot results
+plt.figure(figsize=(10, 6))
+plt.plot(true_rescaled, label='True Values')
+plt.plot(predictions_rescaled, label='Predicted Values', alpha=0.7)
+plt.legend()
+plt.show()
+```
+
+## Music Generation
+
+## Music Generation
+
+The goal is to create an LSTM architecture to create a piece of music. We will train the LSTM on some example snippets of music. The trained model will then be able to generate a new piece of music.
+
+We will first need to write some helper functions to get this to work.
+
+## MIDI Data
+
+We will use the package [pretty_midi](https://github.com/craffel/pretty-midi). This package will allow us to convert MIDI (Musical Instrument Digital Interface) files into NumPy arrays. MIDI files are a common format for music data. They contain information about the notes, timing, and instruments used in a piece of music. We can then use the NumPy arrays to train our LSTM.
+
+Below is the code that will convert our MIDI file to a NumPy array.
+
+```{python}
+#| code-fold: false
+import pretty_midi
+import numpy as np
+
+def midi_to_notes(midi_file):
+    midi_data = pretty_midi.PrettyMIDI(midi_file)
+    notes = []
+    for instrument in midi_data.instruments:
+        if not instrument.is_drum:
+            for note in instrument.notes:
+                notes.append([note.start, note.end, note.pitch, note.velocity])
+    return np.array(notes)
+```
+
+## Maestro dataset
+
+The [Maestro](https://magenta.tensorflow.org/datasets/maestro#v300) dataset is a dataset composed of about 200 hours of virtuosic piano performances. I took a small sample from one of these files to train our LSTM and generate some new music.
+
+```{python}
+#| code-fold: false
+midi_files = [
+    'music/music_sample_1.mid',
+    'music/music_sample_2.mid',
+    'music/music_sample_3.mid']
+all_notes = []
+for midi_file in midi_files:
+  notes = midi_to_notes(midi_file)
+  all_notes.append(notes)
+all_notes = np.concatenate(all_notes)
+```
+
+## Sequences of Notes
+
+Convert the note sequences into a format suitable for training the LSTM model. This means we create sequences of notes and corresponding targets.
```

Comment:
Can we illustrate this somehow? So midi is describing the notes, kind of like PDF portable document format, and what we want are time based samples, right?

Reply:
Added an illustration to show how to break up the data into X and y.